SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift

```diff
@@ -1010,8 +1010,7 @@ func isCastSupportedInEmbeddedSwift(from sourceType: Type,
     return false
   }
 
-  // Tuple?
-  if !destType.isStruct && !destType.isClass && !destType.isEnum {
+  if !destType.isStruct && !destType.isClass && !destType.isEnum && !destType.isTuple {
```

Comment:
Maybe add a comment listing which types are _not_ supported here? Function types? Also why aren't they supported when tuples are supported? Tuples can have arbitrary element types, including function types, in them.